Save runtimes in configmaps

rakesh-garimella opened the pull request:
**Issue Ref**: [Issue number related to this PR or None]
 
**Description**: 

[PR Description]

**TODOs**:
 - [ ] Ready to review
 - [ ] Automated Tests
 - [ ] Docs

rakesh-garimella committed:
Working prototype

rakesh-garimella committed:
Added kubeless-config yaml

rakesh-garimella committed:
Cleanup

rakesh-garimella committed:
Working after cleanup

rakesh-garimella committed:
Modify the client to read the configmap

 - The configmap needs to be read by the client

rakesh-garimella committed:
Made changes based on comments

 - The langruntimes is now part of the controller
   data structure so that it can be passed on to
   k8sutils.go

- The langruntimes reads the configmap on the
  start of the controller.

- Currently monitoring of the configmap is missing.
  So if the configmap is changed the controller needs
  to be restarted.

- The kubeless binary has been updated to read the
  configmap everytime the kubeless binary is used to
  deploy/update the function

- Updated the unit tests

rakesh-garimella committed:
Dont repeat configmap initialization for every test.

sayanh commented on pkg/controller/controller.go:
```diff
@@ -88,13 +90,17 @@ func New(cfg Config, smclient *monitoringv1alpha1.MonitoringV1alpha1Client) *Con
 		},
 	})
 
+	var lr = langruntime.New(cfg.KubeCli, "kubeless", "kubeless-config")
```
Personally, I would love `langruntime` over `lr`

sayanh commented on cmd/kubeless/function/deploy.go:
```diff
@@ -173,7 +177,11 @@ var deployCmd = &cobra.Command{
 }
 
 func init() {
```
General comment: So we have to restart controller to change a runtimes configmap as it is picked up during bootup. Is this the expected behavior?

rakesh-garimella committed:
Updated the documentation for adding custom runtimes.

gchbg commented on cmd/kubeless/function/deploy.go:
```diff
@@ -33,6 +33,11 @@ var deployCmd = &cobra.Command{
 	Long:  `deploy a function to Kubeless`,
 	Run: func(cmd *cobra.Command, args []string) {
+
+		cli := utils.GetClientOutOfCluster()
```
I'm worried that rereading the configmap on every function deploy/update can leave the controller in an inconsistent state if runtimes are changed or removed while there are deployed functions using them. I would guess this is one good reason why kubeless has the "clean functions up on startup" functionality. I would recommend, for now, to have this configmap read only once on controller startup. This will ensure that the behavior of the controller is consistent and also that cleaning up on startup will maintain that consistency.

rakesh-garimella committed:
Update the jsonnet to pass the runtime information as configmaps

Also remove kubeless-config.yaml
Add a new get-server-config flag to print the server configuration
Remove checks from unittest without RBAC. Since controller
fails to read the config file.

andresmgot commented on cmd/kubeless/function/deploy.go:
```diff
@@ -173,7 +177,11 @@ var deployCmd = &cobra.Command{
 
 func init() {
-	deployCmd.Flags().StringP("runtime", "", "", "Specify runtime. Available runtimes are: "+strings.Join(langruntime.GetRuntimes(), ", "))
+	cli := utils.GetClientOutOfCluster()
```
I agree, we can move this info to a new command `kubeless function get-server-config` in which we can show that info and anything else we need in the future.

rakesh-garimella commented on pkg/langruntime/langruntime.go:
```diff
@@ -7,185 +7,215 @@ import (
+}
+
+// ImageSecrets for pulling the image
+type ImageSecrets struct {
```
done

rakesh-garimella commented on pkg/langruntime/langruntime.go:
```diff
@@ -7,185 +7,215 @@ import (
-	node8 := runtimeVersion{version: "8", httpImage: node8Http, pubsubImage: node8Pubsub, initImage: node8Init}
-	nodeVersions = []runtimeVersion{node6, node8}
+	cfgm, err := l.clientset.CoreV1().ConfigMaps("kubeless").Get("kubeless-config", metav1.GetOptions{})
 
```
done

rakesh-garimella commented on pkg/langruntime/langruntime.go:
```diff
@@ -7,185 +7,215 @@ import (
-	ruby24 := runtimeVersion{version: "2.4", httpImage: ruby24Http, pubsubImage: ruby24Pubsub, initImage: ruby24Init}
-	rubyVersions = []runtimeVersion{ruby24}
+	if err != nil {
+		logrus.Fatalf("Unable to get the configmap. %v", err)
```
done

rakesh-garimella commented on pkg/langruntime/langruntime.go:
```diff
@@ -7,185 +7,215 @@ import (
-	dotnetcore2 := runtimeVersion{version: "2.0", httpImage: dotnetcore2Http, pubsubImage: "", initImage: dotnetcore2Init}
-	dotnetcoreVersions = []runtimeVersion{dotnetcore2}
+	err = yaml.Unmarshal([]byte(cfgm.Data["runtime-images"]), &l.AvailableRuntimes)
 
```
done

rakesh-garimella commented on pkg/langruntime/langruntime.go:
```diff
@@ -7,185 +7,215 @@ import (
-		{ID: "ruby", versions: rubyVersions, DepName: "Gemfile", FileNameSuffix: ".rb"},
-		{ID: "dotnetcore", versions: dotnetcoreVersions, DepName: "requirements.xml", FileNameSuffix: ".cs"},
+	if err != nil {
+		logrus.Fatalf("Unable to get the runtime images. %v", err)
```
done

rakesh-garimella committed:
Changes based on PR comments

 - Removed the debug messages from libtest
 - Added unit tests for imagesecrets
 - Moved the common code into langruntimetestutils.go
 - Updated the Readme with information on new switch
   kubeless function.

rakesh-garimella committed:
Made changes based on feed from PR

rakesh-garimella committed:
Resolving merge conflicts

rakesh-garimella committed:
Rebased to latest master

 - Resolved the merge conflicts
 - Reworked langruntimes

rakesh-garimella committed:
Update the documentation to reflect new behaviour

rakesh-garimella committed:
Made changes based on PR Comments

 - Made get-server-config a command/
 - Fixed Readme to reflect changes.
